Baseball models, client, and server

nkreeger opened the pull request:
This is rather large - so getting this started now.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/tensorflow/tfjs-examples/66)
<!-- Reviewable:end -->

dsmilkov commented:
Took a first pass. The 3 major comments I have:
1) Move from TS to JS to make it more accessible
2) Add readme that explains what this demo does, how it works and how to build/develop/test
3) Not blocking: Remove Vue and use plain JS for the UI to simplify the project and make it more accessible

---

Reviewed 19 of 24 files at r1, 5 of 5 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[.gitignore, line 1 at r2](https://reviewable.io/reviews/tensorflow/tfjs-examples/66#-LBkqPJp0RXU6MLBgYIC:-LBkqPJp0RXU6MLBgYID:bcyqwfx) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/345301a8c4faec064e64bc1f2b967b99cfd288be/.gitignore#L1)):*
> ```
> node_modules
> ```

Since we've standardized all the examples, we have a global [README.md](https://github.com/tensorflow/tfjs-examples) that talks how to build and run any of the examples, in addition on an example-specific readme that is much simpler. In this case, since this example is more complex (with server/client architecture), we should definitely have a readme with instructions on how this example is structured, what technologies it uses and how to build and test.

---

*[.gitignore, line 7 at r2](https://reviewable.io/reviews/tensorflow/tfjs-examples/66#-LBkfzZzEdStgS6cTZjq:-LBkfz_-0cwLlAPVMihk:bu5dhkf) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/345301a8c4faec064e64bc1f2b967b99cfd288be/.gitignore#L7)):*
> ```
> *.tgz
> *.pyc
> *_data.json
> ```

Store the downloaded data in baseball/dist/ where all generated files go. This way you don't have to git-ignore the json explicitly here.

---

*[baseball/package.json, line 1 at r2](https://reviewable.io/reviews/tensorflow/tfjs-examples/66#-LBkgOvV7d1DxdaduACi:-LBkgOvV7d1DxdaduACj:bpcu8u8) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/345301a8c4faec064e64bc1f2b967b99cfd288be/baseball/package.json#L1)):*
> ```JSON
> {
> ```

rename the folder to baseball-node to clarify that it is a node demo. This will align with other examples where we use -core and -layers suffix

---

*[baseball/package.json, line 3 at r2](https://reviewable.io/reviews/tensorflow/tfjs-examples/66#-LBkgtow3kzWAByhk9HV:-LBkgtow3kzWAByhk9HW:b-6yb7sy) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/345301a8c4faec064e64bc1f2b967b99cfd288be/baseball/package.json#L3)):*
> ```JSON
> {
>   "name": "tfjs-examples-baseball",
>   "version": "0.1.0",
> ```

might be good to add "private": "true" - an additional signal we won't publish this to npm.

---

*[baseball/package.json, line 5 at r2](https://reviewable.io/reviews/tensorflow/tfjs-examples/66#-LBkguV7DAPeNZeYeeRm:-LBkguV7DAPeNZeYeeRn:b-dvt49w) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/345301a8c4faec064e64bc1f2b967b99cfd288be/baseball/package.json#L5)):*
> ```JSON
>   "version": "0.1.0",
>   "description": "Deep learning models for classifying baseball metrics",
>   "main": "index.js",
> ```

do we need this "main" property? I think you can remove it

---

*[baseball/package.json, line 9 at r2](https://reviewable.io/reviews/tensorflow/tfjs-examples/66#-LBkhAex9dZtrLEjt6-T:-LBkhAex9dZtrLEjt6-U:b-ky9h8h) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/345301a8c4faec064e64bc1f2b967b99cfd288be/baseball/package.json#L9)):*
> ```JSON
>     "build": "tsc",
>     "build-client": "webpack",
>     "link-local": "yalc link",
> ```

Are we pretty close to publishing tfjs-node to npm? Asking since if you can publish to npm before we check in this demo, then we don't have to come back and update this demo. Also we will dogfood the npm and simplify the build process (no need to link-local)

---

*[baseball/package.json, line 26 at r2](https://reviewable.io/reviews/tensorflow/tfjs-examples/66#-LBki7aL4bLxK0NpCwYD:-LBki7aL4bLxK0NpCwYE:bxducu0) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/345301a8c4faec064e64bc1f2b967b99cfd288be/baseball/package.json#L26)):*
> ```JSON
>     "clang-format": "^1.2.3",
>     "containers.js": "^0.0.6",
>     "css-loader": "^0.28.11",
> ```

replace all "^..." with "~..." to be a little more strict with the versioning. This helps reduce breakages in the fast moving npm world

---

*[baseball/package.json, line 43 at r2](https://reviewable.io/reviews/tensorflow/tfjs-examples/66#-LBkhyPVAyUr5sG5oR_0:-LBkhyPVAyUr5sG5oR_1:b7q3d6d) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/345301a8c4faec064e64bc1f2b967b99cfd288be/baseball/package.json#L43)):*
> ```JSON
>     "yalc": "^1.0.0-pre.22"
>   },
>   "dependencies": {}
> ```

move tfjs, tfjs-node and other things that are used at run-time from devDependencies down here

---

*[baseball/tsconfig.json, line 2 at r2](https://reviewable.io/reviews/tensorflow/tfjs-examples/66#-LBkiVe43M6CyTSB-NQF:-LBkiVe43M6CyTSB-NQG:bf6xg2i) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/345301a8c4faec064e64bc1f2b967b99cfd288be/baseball/tsconfig.json#L2)):*
> ```JSON
> {
>   "compilerOptions": {
> ```

All of the other `tfjs-examples` are javascript to make them more accessible. This is based on a consistent feedback we've gotten from the community. Typescript people don't mind JS, JS people mind TS. Hopefully it won't be too hard to remove all ts annotations and turn the files into modern (es7) js.

---

*[baseball/webpack.config.js, line 1 at r2](https://reviewable.io/reviews/tensorflow/tfjs-examples/66#-LBkjUmQ6JVxzKv6GV-_:-LBkjUmQ6JVxzKv6GV-a:b-z6llk3) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/345301a8c4faec064e64bc1f2b967b99cfd288be/baseball/webpack.config.js#L1)):*
> ```JavaScript
> var path = require('path')
> ```

add license

---

*[baseball/webpack.config.js, line 14 at r2](https://reviewable.io/reviews/tensorflow/tfjs-examples/66#-LBklfXt-K5ltAwiopEH:-LBklfXt-K5ltAwiopEI:bo7nzgf) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/345301a8c4faec064e64bc1f2b967b99cfd288be/baseball/webpack.config.js#L14)):*
> ```JavaScript
>     rules: [
>       {
>         test: /\.vue$/,
> ```

Again based on community feedback. Not in this PR, but definitely worth filing a bug: To make the demo simpler and more accessible and to keep the focus on the library itself, we should remove vue and just use a plain JS for the UI.

---

*[baseball/src/abstract-pitch-model.ts, line 26 at r2](https://reviewable.io/reviews/tensorflow/tfjs-examples/66#-LBknz766LNMurqDi7nZ:-LBknz766LNMurqDi7n_:b-ac2hu3) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/345301a8c4faec064e64bc1f2b967b99cfd288be/baseball/src/abstract-pitch-model.ts#L26)):*
> ```XML
>  * Abstract base class for defining Pitch ML models.
>  */
> export abstract class PitchModel {
> ```

noticed it's an abstract class without any abstract methods, whick makes it a bit hard to identify what the job of the subclass is.

---

*[baseball/src/pitch-data.ts, line 19 at r2](https://reviewable.io/reviews/tensorflow/tfjs-examples/66#-LBkoTLAC14T-a2XDERb:-LBkoTLAC14T-a2XDERc:beo5qgx) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/345301a8c4faec064e64bc1f2b967b99cfd288be/baseball/src/pitch-data.ts#L19)):*
> ```XML
> 
> import * as tf from '@tensorflow/tfjs';
> import {sizeFromShape} from '@tensorflow/tfjs-core/dist/util';
> ```

Use the public API `tf.util.sizeFromShape` instead, since this import approach is error prone to tfjs-core moving files around.

---

*[baseball/src/utils.ts, line 22 at r2](https://reviewable.io/reviews/tensorflow/tfjs-examples/66#-LBkp2oq7UD_w36mg7Bj:-LBkp2oq7UD_w36mg7Bk:bdi0as9) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/345301a8c4faec064e64bc1f2b967b99cfd288be/baseball/src/utils.ts#L22)):*
> ```XML
>  */
> // tslint:disable-next-line:no-any
> export function shuffle(array: any[]): any[] {
> ```

use `tf.util.shuffle`

---

*[baseball/src/train/train-pitch-type-model.ts, line 24 at r2](https://reviewable.io/reviews/tensorflow/tfjs-examples/66#-LBkq5BCAZZBtEoy2eE2:-LBkq5BCAZZBtEoy2eE3:b-7djdj5) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/345301a8c4faec064e64bc1f2b967b99cfd288be/baseball/src/train/train-pitch-type-model.ts#L24)):*
> ```XML
> 
> bindTensorFlowBackend();
> ```

Is this used only for testing? In general, what's the purpose of the `src/train/` folder? Perhaps we can remove it now that the demo is stable.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-examples/66#-:-LBkr2Hf-CMkWAEXKxfL:b62riru)*
<!-- Sent from Reviewable.io -->

nkreeger commented:
Review status: 0 of 25 files reviewed at latest revision, 16 unresolved discussions.

---

*[.gitignore, line 7 at r2](https://reviewable.io/reviews/tensorflow/tfjs-examples/66#-LBkfzZzEdStgS6cTZjq:-LBlDEzF5osYL4hp9poZ:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/345301a8c4faec064e64bc1f2b967b99cfd288be/.gitignore#L7)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

Store the downloaded data in baseball/dist/ where all generated files go. This way you don't have to git-ignore the json explicitly here.
</blockquote></details>

Done.

---

*[baseball/package.json, line 1 at r2](https://reviewable.io/reviews/tensorflow/tfjs-examples/66#-LBkgOvV7d1DxdaduACi:-LBlDeA6DPSsYs3eU0Yo:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/345301a8c4faec064e64bc1f2b967b99cfd288be/baseball/package.json#L1)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

rename the folder to baseball-node to clarify that it is a node demo. This will align with other examples where we use -core and -layers suffix
</blockquote></details>

Done.

---

*[baseball/package.json, line 3 at r2](https://reviewable.io/reviews/tensorflow/tfjs-examples/66#-LBkgtow3kzWAByhk9HV:-LBlDkLG7q2JvBtv3xF1:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/345301a8c4faec064e64bc1f2b967b99cfd288be/baseball/package.json#L3)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

might be good to add "private": "true" - an additional signal we won't publish this to npm.
</blockquote></details>

Done.

---

*[baseball/package.json, line 5 at r2](https://reviewable.io/reviews/tensorflow/tfjs-examples/66#-LBkguV7DAPeNZeYeeRm:-LBlDlPp1YvOutfabqTl:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/345301a8c4faec064e64bc1f2b967b99cfd288be/baseball/package.json#L5)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

do we need this "main" property? I think you can remove it
</blockquote></details>

Done.

---

*[baseball/package.json, line 9 at r2](https://reviewable.io/reviews/tensorflow/tfjs-examples/66#-LBkhAex9dZtrLEjt6-T:-LBlDmFrAdWFrsOOFpoJ:b-ggyacr) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/345301a8c4faec064e64bc1f2b967b99cfd288be/baseball/package.json#L9)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

Are we pretty close to publishing tfjs-node to npm? Asking since if you can publish to npm before we check in this demo, then we don't have to come back and update this demo. Also we will dogfood the npm and simplify the build process (no need to link-local)
</blockquote></details>

Still waiting on internal approvals. Probably going to be early next week.

---

*[baseball/package.json, line 43 at r2](https://reviewable.io/reviews/tensorflow/tfjs-examples/66#-LBkhyPVAyUr5sG5oR_0:-LBlE4l3-ngWEO_8pwPo:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/345301a8c4faec064e64bc1f2b967b99cfd288be/baseball/package.json#L43)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

move tfjs, tfjs-node and other things that are used at run-time from devDependencies down here
</blockquote></details>

Done.

---

*[baseball/webpack.config.js, line 1 at r2](https://reviewable.io/reviews/tensorflow/tfjs-examples/66#-LBkjUmQ6JVxzKv6GV-_:-LBliu8b8KJm9Sui-L3D:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/345301a8c4faec064e64bc1f2b967b99cfd288be/baseball/webpack.config.js#L1)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

add license
</blockquote></details>

Done.

---

*[baseball/src/abstract-pitch-model.ts, line 26 at r2](https://reviewable.io/reviews/tensorflow/tfjs-examples/66#-LBknz766LNMurqDi7nZ:-LBliyei585DuPwe7lkZ:b-yz8gcv) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/345301a8c4faec064e64bc1f2b967b99cfd288be/baseball/src/abstract-pitch-model.ts#L26)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

noticed it's an abstract class without any abstract methods, whick makes it a bit hard to identify what the job of the subclass is.
</blockquote></details>

I took a pass, renamed this and added some comments. Mostly, training is abstracted into this class - models simply subclass and just define the training fields, label, and prediction method.

---

*[baseball/src/pitch-data.ts, line 19 at r2](https://reviewable.io/reviews/tensorflow/tfjs-examples/66#-LBkoTLAC14T-a2XDERb:-LBljAMHA13l5vqxjXUH:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/345301a8c4faec064e64bc1f2b967b99cfd288be/baseball/src/pitch-data.ts#L19)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

Use the public API `tf.util.sizeFromShape` instead, since this import approach is error prone to tfjs-core moving files around.
</blockquote></details>

Done.

---

*[baseball/src/utils.ts, line 22 at r2](https://reviewable.io/reviews/tensorflow/tfjs-examples/66#-LBkp2oq7UD_w36mg7Bj:-LBljH8wAIGGvMfzCYr1:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/345301a8c4faec064e64bc1f2b967b99cfd288be/baseball/src/utils.ts#L22)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

use `tf.util.shuffle`
</blockquote></details>

Done.

---

*[baseball/src/train/train-pitch-type-model.ts, line 24 at r2](https://reviewable.io/reviews/tensorflow/tfjs-examples/66#-LBkq5BCAZZBtEoy2eE2:-LBlEJ1oAPzzIeB8RTsA:b-ng72r1) ([raw file](https://github.com/tensorflow/tfjs-examples/blob/345301a8c4faec064e64bc1f2b967b99cfd288be/baseball/src/train/train-pitch-type-model.ts#L24)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

Is this used only for testing? In general, what's the purpose of the `src/train/` folder? Perhaps we can remove it now that the demo is stable.
</blockquote></details>

These are sample scripts to demonstrate training w/o all the client/server infrastructure. I was going to put these in a readme.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-examples/66)*
<!-- Sent from Reviewable.io -->

dsmilkov commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Reviewed 25 of 56 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-examples/66#-:-LBnYubK2X_VeQvZR9mr:b-jbsf8z)*
<!-- Sent from Reviewable.io -->
